doc/decisions/constructor_functions.md

```diff
@@ -0,0 +1,85 @@
+# Constructor Functions
+
+## Problem
+
+The constructor functions that `libelektra-core` defines for creating Keys and KeySets must be possible to implement in other languages without complicated workarounds.
+In particular, another implementation should not need to resort to adding a small amount of C to "bridge the gap".
+
+## Constraints
+
+- Constructor functions should use as little resources as feasible.
+  In particular, the number of `malloc`s and `memcpy` should be kept in check.
+  Temporary allocations during key creation should be avoided.
+- Using the escaped name in the constructor function should be avoided, since the unescaping process can be expensive.
+
+## Assumptions
+
+- The constructor functions from `libelektra-core` will not or only very rarely be used directly by users.
+- For all languages supported by Elektra (including C), there is secondary library on top of `libelektra-core` that provides more idiomatic APIs.
+- Creating a key with a name, no value and no metadata and then calling `keySetValue`/`keySetMeta` does not create significant overhead compared to creating the key with value and metadata directly.
```

Comment:
This is the key of the whole decision. I don't think we should base this on assumptions but actually benchmark keyNew if avoiding calling `keySetValue`/`keySetMeta` is a relevant topic or not.

Simply leaving it as is is also totally possible. In any case it is good that we now have this decision.

Reply:
To clarify, the assumption says:

There is no significant overhead when switching from
```c
Key * k = keyNew ("/foo", KEY_VALUE, "value", KEY_META, "meta:/type", "string", KEY_META, "meta:/bar", "baz", KEY_END);
```

to

```c
Key * k =keyNew ("/foo", KEY_END);
keySetString (k, "value");
keySetMeta (k, "meta:/type", "string");
keySetMeta (k, "meta:/bar", "baz");
```

This is currently true, because `keyNew` internally just call `keySetString`/`keySetMeta`. (*) The assumption is just that, if `keyNew` were to directly set the value and directly construct the correct `struct _KeySet`, there still wouldn't be a significant difference. Since there are no optimizations that `keyNew` could do, that couldn't be done by `keySetValue`/`keySetMeta`, I think this is a fair assumption.

Additionally, the assumption is actually only semi-relevant to the decision. In fact, the actual decision, makes the assumption less relevant. Since `keyCreate` could still use the private API and directly construct the `struct _Key`, any optimization could simply be implemented there.

---

(*) In theory, there could be (significant) difference between 1 call and 4 calls to dynamically linked symbols. However, if that change is significant, the whole idea of a minimal API becomes question able. If the API is minimal, you always need more calls to do complicated stuff, compared to an API that has functions for everything.